src/Components/Server/src/Circuits/CircuitRegistry.cs

```diff
@@ -104,44 +107,60 @@ public virtual Task DisconnectAsync(CircuitHost circuitHost, string connectionId
 
         protected virtual bool DisconnectCore(CircuitHost circuitHost, string connectionId)
         {
-            if (!ConnectedCircuits.TryGetValue(circuitHost.CircuitId, out circuitHost))
+            var circuitId = circuitHost.CircuitId;
+            if (!ConnectedCircuits.TryGetValue(circuitId, out circuitHost))
             {
+                Log.CircuitNotActive(_logger, circuitId);
+
                 // Guard: The circuit might already have been marked as inactive.
                 return false;
             }
 
             if (!string.Equals(circuitHost.Client.ConnectionId, connectionId, StringComparison.Ordinal))
             {
+                Log.CircuitConnectedToDifferentConnection(_logger, circuitId, circuitHost.Client.ConnectionId);
+
                 // The circuit is associated with a different connection. One way this could happen is when
                 // the client reconnects with a new connection before the OnDisconnect for the older
                 // connection is executed. Do nothing
                 return false;
             }
 
-            var result = ConnectedCircuits.TryRemove(circuitHost.CircuitId, out circuitHost);
+            var result = ConnectedCircuits.TryRemove(circuitId, out circuitHost);
             Debug.Assert(result, "This operation operates inside of a lock. We expect the previously inspected value to be still here.");
 
             circuitHost.Client.SetDisconnected();
             RegisterDisconnectedCircuit(circuitHost);
+
+            Log.CircuitMarkedDisconnected(_logger, circuitId);
+
             return true;
         }
 
         public void RegisterDisconnectedCircuit(CircuitHost circuitHost)
         {
+            var cancellationTokenSource = new CancellationTokenSource(_options.DisconnectedCircuitRetentionPeriod);
```

Comment:
Will this create 1 timer per circuit? Is that something we should care about? Should this instead be handled by something that executes at regular intervals and clears circuits past some expiration time?

Reply:
> so it should be fairly efficient.

😄 Famous last words. While it is efficient. If most of your objects have the same interval then it might be better to have a single recurring timer that loops over all of your circuits. SignalR tries really hard to avoid making too many timers.

Reply:
Let's profile this first and we can figure it out